src/algebra/quaternion_basis.lean

```diff
@@ -0,0 +1,166 @@
+/-
+Copyright (c) 2021 Eric Wieser. All rights reserved.
+Released under MIT license as described in the file LICENSE.
+Authors: Eric Wieser
+-/
+import algebra.quaternion
+import tactic.ring
+
+/-!
+# Basis on a quaternion-like algebra
+
+## Main definitions
+
+* `quaternion_algebra.basis A c₁ c₂`: a basis for a subspace of an `R`-algebra `A` that has the
+  same algebra structure as `ℍ[R,c₁,c₂]`.
+* `quaternion_algebra.basis.self R`: the canonical basis for `ℍ[R,c₁,c₂]`.
+* `quaternion_algebra.basis.comp_hom b f`: transform a basis `b` by an alg_hom `f`.
+* `quaternion_algebra.lift`: Define an `alg_hom` out of `ℍ[R,c₁,c₂]` by its action on the basis
+  elements `i`, `j`, and `k`. In essence, this is a universal property. Analogous to `complex.lift`,
+  but takes a bundled `quaternion_algebra.basis` instead of just a `subtype` as the amount of
+  data / proves is non-negligeable.
+-/
+
+open_locale quaternion
+namespace quaternion_algebra
+
+
+/-- A quaternion basis contains the information sufficient to show that a subalgebra of `A`
+is compatible with `ℍ[R,c₁,c₂]`.
```

Comment:
```suggestion
is isomorphic to `ℍ[R,c₁,c₂]`.
```
is that what you mean? Or do you mean something else? I guess it's either isomorphic to the quaternion algebra, or `A` is the zero ring?

Reply:
I mean something else. Essentially, this structure is "sufficient and necessary for an alg_hom to exist from `ℍ[R,c₁,c₂]` to `A`. This only represents an isomorphism to the subalgebra if the alg_hom is injective, which I suspect does not have to be the case.

Reply:
"has a surjective morphism from" is perhaps more accurate.